executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,686 @@
+			}
+		}
+		dLookUpKeys = append(dLookUpKeys, dLookUpKey)
+	}
+
+	return dLookUpKeys, nil
+}
+
+func (imw *innerMergeWorker) constructDatumLookupKey(task *lookUpMergeJoinTask, rowIdx int) (*indexJoinLookUpContent, error) {
+	if task.outerMatch != nil && !task.outerMatch[rowIdx] {
+		return nil, nil
+	}
+	outerRow := task.outerResult.GetRow(rowIdx)
+	sc := imw.ctx.GetSessionVars().StmtCtx
+	keyLen := len(imw.keyCols)
+	dLookupKey := make([]types.Datum, 0, keyLen)
+	for i, keyCol := range imw.outerMergeCtx.keyCols {
+		outerValue := outerRow.GetDatum(keyCol, imw.outerMergeCtx.rowTypes[keyCol])
+		// Join-on-condition can be promised to be equal-condition in
+		// IndexNestedLoopJoin, thus the filter will always be false if
+		// outerValue is null, and we don't need to lookup it.
+		if outerValue.IsNull() {
+			return nil, nil
+		}
+		innerColType := imw.rowTypes[imw.keyCols[i]]
+		innerValue, err := outerValue.ConvertTo(sc, innerColType)
+		if err != nil {
+			// If the converted outerValue overflows, we don't need to lookup it.
+			if terror.ErrorEqual(err, types.ErrOverflow) {
+				return nil, nil
+			}
+			return nil, err
+		}
+		cmp, err := outerValue.CompareDatum(sc, &innerValue)
+		if err != nil {
+			return nil, err
+		}
+		if cmp != 0 {
+			// If the converted outerValue is not equal to the origin outerValue, we don't need to lookup it.
+			return nil, nil
+		}
+		dLookupKey = append(dLookupKey, innerValue)
+	}
+	return &indexJoinLookUpContent{keys: dLookupKey, row: task.outerResult.GetRow(rowIdx)}, nil
+}
+
+func (imw *innerMergeWorker) dedupDatumLookUpKeys(lookUpContents []*indexJoinLookUpContent) []*indexJoinLookUpContent {
+	if len(lookUpContents) < 2 {
+		return lookUpContents
+	}
+	sc := imw.ctx.GetSessionVars().StmtCtx
+	deDupedLookUpContents := lookUpContents[:1]
+	for i := 1; i < len(lookUpContents); i++ {
+		cmp := compareRow(sc, lookUpContents[i].keys, lookUpContents[i-1].keys)
+		if cmp != 0 || (imw.nextColCompareFilters != nil && imw.nextColCompareFilters.CompareRow(lookUpContents[i].row, lookUpContents[i-1].row) != 0) {
+			deDupedLookUpContents = append(deDupedLookUpContents, lookUpContents[i])
+		}
+	}
+	// If the order requires descent, the deDupedLookUpContents is keep descent order before.
+	// So at the end, we should generate the ascent deDupedLookUpContents to build the correct range for inner read.
+	if !imw.outerMergeCtx.needOuterSort && imw.desc {
+		lenKeys := len(deDupedLookUpContents)
+		for i := 0; i < lenKeys/2; i++ {
+			deDupedLookUpContents[i], deDupedLookUpContents[lenKeys-i-1] = deDupedLookUpContents[lenKeys-i-1], deDupedLookUpContents[i]
+		}
+	}
+	return deDupedLookUpContents
+}
+
+func (imw *innerMergeWorker) fetchInnerResult(ctx context.Context, task *lookUpMergeJoinTask) (beginRow chunk.Row, err error) {
+	task.innerResult = chunk.NewChunkWithCapacity(retTypes(imw.innerExec), imw.ctx.GetSessionVars().MaxChunkSize)
+	err = Next(ctx, imw.innerExec, task.innerResult)
+	if task.innerResult != nil {
```

Comment:
1. This check will always be true since we've built a new chunk for task.innerResult.
2. Why do we build a new chunk for task.innerResult but not use task.innerResult.Reset()

Reply:
Using `task.innerResult.Reset()` will cause wrong results. Because `sameKeyRows` may contains several `innerResult` chunks. If we reset it, the content of `sameKeyRows` is going to be wrong.